blocks/algorithms/__init__.py

```diff
@@ -0,0 +1,239 @@
+
+    Very often the goal of training is to minimize the expected value of a
+    Theano expression. Batch processing in this cases typically consists of
+    running a (or a few) Theano functions.
+    :class:`DifferentiableCostMinimizer` is the base class for such
+    algorithms.
+
+    Parameters
+    ----------
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned. If ``None``, all shared variables of
+        `cost` computation graph will be considered parameters.
+
+    Attributes
+    ----------
+    updates : list of (shared variable, Theano expression) tuples
+        Updates to be done for every batch. It is required that the
+        updates are done using the old values of optimized parameters.
+    cost : Theano variable
+        The objective to be minimized.
+    params : list of Theano shared variables
+        The parameters to be tuned.
+
+    Notes
+    -----
+        Changing `updates` attribute or calling `add_updates` after
+        the `initialize` method is called will have no effect.
+
+    .. todo::
+
+        Some shared variables are not parameters (e.g. those created by
+        random streams)
+
+    .. todo::
+
+        Due to a rather premature status of the :class:`ComputationGraph`
+        class the parameter used only inside scans are not fetched
+        currently.
+
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, cost, params=None):
+        self.cost = cost
+        self.params = (params if params
+                       else ComputationGraph(cost).get_shared_variables())
+        self._cost_computation_graph = ComputationGraph(self.cost)
+        self._updates = []
+
+    @property
+    def inputs(self):
+        """Return inputs of the cost computation graph.
+
+        Returns
+        -------
+            list of Theano variables
+
+        """
+        return self._cost_computation_graph.inputs
+
+    @property
+    def updates(self):
+        return self._updates
+
+    @updates.setter
+    def updates(self, value):
+        self._updates = value
+
+    def add_updates(self, updates):
+        """Add updates to the training process.
+
+        The updates will be done _before_ the parameters are changed.
+
+        Parameters
+        ----------
+        updates : list of tuples or :class:`OrderedDict`
+            The updates to add.
+
+        """
+        if isinstance(updates, OrderedDict):
+            updates = list(updates.items())
+        assert isinstance(updates, list)
+        self.updates.extend(updates)
+
+
+class GradientDescent(DifferentiableCostMinimizer):
+    """A base class for all gradient descent algorithms.
+
+    By "gradient descent" we mean a training algorithm of the following
+    form:
+
+    .. code-block::  python
+
+        for batch in data:
+            for param in params:
+                param += step.compute(grad_wr_param_on_batch)
```

Comment:
Ah, no, code-wise I think it makes sense (since the step is in the negative gradient direction). I just read this docstring as saying "Gradient descent is when you do `param += grad_wr_param_on_batch`", thinking that `compute` meant "compute the gradient". It makes sense if you know the code, otherwise it might be a bit confusing, but it's minor.